src/gorilla.c

```diff
@@ -0,0 +1,261 @@
+/*
+* Copyright 2018-2019 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include <assert.h>
+
+#include "gorilla.h"
+
+
+#define _64BIT 64
+#define DOUBLE_LEADING 5
+#define DOUBLE_BLOCK_SIZE 6
+#define DOUBLE_BLOCK_ADJUST 1
+
+#define CHECKSPACE(chunk, x)    if (!isSpaceAvailable(chunk, x)) { \
+                                    return CC_ERR; }
+
+#define  LeadingZeros64(x)  __builtin_clzll(x)
+#define TrailingZeros64(x)  __builtin_ctzll(x)
+
+// Clear all bits from position `bits`
+static u_int64_t clearBits(u_int64_t data, u_int64_t bits) {
+    return data & ((1LL << bits) - 1);
+}
+
+// Converts `x`, an int64, to binary representation with length `l`
+static u_int64_t int2bin(int64_t x, u_int8_t l) {
+    u_int64_t negative = x < 0 ? 1 : 0;
+    u_int64_t rv = clearBits(x, l - 1);
+    rv |= (negative << (l - 1));
+    return rv;
+}
+
+// Converts `x`, a binary of length `l`, into an int64
+static int64_t bin2int(u_int64_t x, u_int8_t l) {
+    bool negative = x & (1LL << (l - 1));
+    int64_t rv = x;
+    if (negative) {
+        rv -= 1LL << l;
+    }
+    return rv;
+}
+
+// Return `true` if byte is off
+static bool isBitOff(u_int64_t *x, u_int64_t idx) {
+    return !(x[idx / 64LL] & (1LL << (idx % 64)));
+}
+
+// Append `dataLen` bits from `data` into `arr` at position `idx`
+static void appendBits(u_int64_t *arr, u_int64_t *idx, u_int64_t data, u_int8_t dataLen) {
+    u_int8_t idx64 = *idx % _64BIT;
+    u_int8_t available = _64BIT - idx64;
+    u_int64_t *runner = &arr[*idx / _64BIT];
+
+    if(available >= dataLen) {
+        *runner |= clearBits(data, dataLen) << idx64;
+    } else {
+        *runner |= data << idx64;
+        *(++runner) |= clearBits(data, dataLen) >> available;
+    }
+    *idx += dataLen;
+}
+
+// Read `dataLen` bits from `arr` at position `idx`
+static u_int64_t readBits(u_int64_t *arr, u_int64_t *idx, u_int8_t dataLen) {
+    u_int64_t rv = 0;
+    u_int8_t idx64 = *idx % _64BIT;
+    u_int8_t available = _64BIT - idx64;
+    u_int64_t *runner = &arr[*idx / _64BIT];
+
+    if(available >= dataLen) {
+        rv = clearBits(*runner >> idx64, dataLen);
+    } else {
+        u_int8_t left = dataLen - available;
+        rv = clearBits(*runner >> idx64, available);
```

Comment:
I might be wrong, but should the first x bits be shifted by `left` and the rest should be just be used with binary or?

Reply:
@danni-m 
I don't think I follow.
Can you write it as a code?